activerecord/lib/active_record/connection_adapters/abstract/connection_handler.rb

```diff
@@ -103,30 +103,32 @@ def connection_pool_list(role = ActiveRecord::Base.current_role)
 
       def establish_connection(config, owner_name: Base, role: ActiveRecord::Base.current_role, shard: Base.current_shard)
         owner_name = StringConnectionOwner.new(config.to_s) if config.is_a?(Symbol)
-
         pool_config = resolve_pool_config(config, owner_name, role, shard)
         db_config = pool_config.db_config
 
-        # Protects the connection named `ActiveRecord::Base` from being removed
-        # if the user calls `establish_connection :primary`.
-        if owner_to_pool_manager.key?(pool_config.connection_specification_name)
+        # If there is an existing pool with the same values as the pool_config
+        # don't remove the connection. Connections should only be removed if we are
+        # establishing a connection on a class that is already connected to a different
+        # configuration.
+        pool = retrieve_connection_pool(pool_config.connection_specification_name, role: role, shard: shard)
+
+        unless pool && pool.db_config == db_config
           remove_connection_pool(pool_config.connection_specification_name, role: role, shard: shard)
-        end
 
-        message_bus = ActiveSupport::Notifications.instrumenter
-        payload = {}
-        if pool_config
-          payload[:spec_name] = pool_config.connection_specification_name
-          payload[:shard] = shard
-          payload[:config] = db_config.configuration_hash
+          owner_to_pool_manager[pool_config.connection_specification_name] ||= PoolManager.new
+          pool_manager = get_pool_manager(pool_config.connection_specification_name)
+          pool_manager.set_pool_config(role, shard, pool_config)
+          pool = pool_config.pool
         end
 
-        owner_to_pool_manager[pool_config.connection_specification_name] ||= PoolManager.new
-        pool_manager = get_pool_manager(pool_config.connection_specification_name)
-        pool_manager.set_pool_config(role, shard, pool_config)
+        payload = {
+          spec_name: pool_config.connection_specification_name,
+          shard: shard,
+          config: db_config.configuration_hash
+        }
 
-        message_bus.instrument("!connection.active_record", payload) do
-          pool_config.pool
+        ActiveSupport::Notifications.instrumenter.instrument("!connection.active_record", payload) do
```

Comment:
It's unlikely to matter either way in practice (mostly because the whole point of being okay with changing this is that we don't think real people are doing widespread same-config re-establishment in the first place), but I feel like we should only do this if the call had an effect.

Reply:
Yea I was on the fence about this. My first implementation didn't instrument but then I wasn't sure if someone still wanted to know that establish_connection was called. Removing it makes an existing pool 4x faster than the prior behavior.

Reply:
I'm leaving this in place for now. I made a refactor that skipped and included other changes when we have a pool, but that refactor resulted in a mysql2 test run with a bunch of flakes I couldn't reproduce locally. I'm going to do that refactor in a followup PR so I can more easily isolate what's going on there from the other fixes here.

Reply:
Our own code inside fixtures consume this instrumentation, but this instrumentation is private (that is why it starts with `!`) so people should not be using.

Reply:
Oh good to know, I want to change the keys in a followup PR anyway so I'll hopefully get rid of it then.